### ISSUE TITLE
Question: How are small files stored?

### ISSUE THREAD

**Issue by hlee-va** [lizardfs/lizardfs#676]
Searched but didn't find any answer:
How are small files (i.e. smaller than the chunk sizes which is 64MiB) stored? Will small files be combined together and saved into one chunk files?

We are storing hundreds of millions of small files (from 100 Bytes to 30KB) in an NFS and we are suffering the "lots of small files" problem right now. Is Lizardfs working well in this case?

**Comment by 4Dolio** [comment]
It stores each in it's own chunk file which is toughly the size of the original data.

I am not sure if there are any optimizations on the road map.  But i think we have talked about ideas like combining files as you suggested.

I have resorted to creating a single large spars file in LFS and then loopback formatting that in order to more efficiently store huge quantities of small files.  Another trick i have used is to back iSCSI with LFS which is similar to my loopback workaround.

The problem with many small files in LFS is more so a case of it makes the metadata set massive on the master nodes, as opposed to having too many chunk files.. but too many chunk files could also be a problem, I suppose.

**Comment by jkiebzak** [comment]
How stable is your filesystem on top of the loop back device? Also, may I
ask how big of a filesystem you are using?

>

**Comment by 4Dolio** [comment]
Very stable.  Only gives me trouble when I do something wrong and malicious.  Of course adding such layers may reduce performance.  There are trade offs obviously, pros and cons, but it is truly amazing just how versitile it can be.

I run each component on bare metal, VMs, Arm micro computers and store chunks on JBODs, SSDs, M2s, virtual disks, and even CIFS mounts.  I have clusters on 3 continents and probably a total near a half PB down to my home cluster on 8TB usb3 and whatever junk disks I have lying around.

Once you get confertable with it you can make it work almost anywhere...  And I trust it completely with my data even when I do not trust some of my hardware..

**Comment by 4Dolio** [comment]
... I wish I could somehow port it to android and run it on a cluster of old crummy phones..

**Comment by jkiebzak** [comment]
I meant, how big is the loop back file system? What did you format the loop
back with? XFS?

**Comment by 4Dolio** [comment]
Loopbacks i have used are up to about 4TB.  I use zfs or ext4 so far, i would stick to a journaling FS of some sort. Not sure if XFS does journaling but it should work all the same.

The only combination that has not worked is:  bare metal ( Xen LVM on MainBoard Raid5 of 4 SSDs ( local Xen SR ( formatted Chunk Server VM Disk with ZFS ( iSCSI VM mounding self for FILE backing with IET ( Xen mounting iSCSI for SR ( Xen VM disks using this iSCSI SR [observed data corruption in VMs mostly with ZFS but also EXT] ) ) ) ) ) ).  If I use EXT4 to format the VM CS Disk instead of ZFS the mysterious data corruptions stopped..

Yes that is a lot of abstraction layers, i know it sounds a little insane. But it works well with EXT4 on the VMs doing CS+Master+iSCSI layer.

**Comment by pkali** [comment]
Important factor to consider is RAM usage per filesystem object. It is roughly 150 bytes per object (file/folder), so 1 billion (10**9) files will take about 150 GiB RAM.
RAM usage can be minimised with use of BerkeleyDB for file names (it is option in - then filenames are stored on disk and cached to RAM.

Another issue for all distributed file systems is latency. I have just written a short explanation for our mailing list, copying it here:
==========================
Small files are one of the hardest cases for distributed file systems in general as the current architecture has its bottlenecks in terms of network latency. 

Simplistically the file read time could be estimated as
    t(f) = L + size(f)/speed 
where
   t(f) -- read time for file `f`
   L -- latency (network, FS, OS), usually in order of miliseconds
   size(f) -- file size
   speed -- raw transfer speed 

For very small files L, while relatively constant, is usually much larger than size(f)/speed, what makes the overall time high and accumulated transfer rate much lower.
This can be alleviated for classes of cases, but requires additional development in the predictor (read-ahead cache) module that is now reading ahead within a given file. 

Example on realistic numbers (simplified) :
for ~10KiB file and low latency network and 500[MiB/s] raw transfer speed.

t(f) = 2[ms] + 0.01[MiB]/500[MiB/s] = 2[ms] + 0.00002[s] = 2.02[ms]
While transfering 100 such files we have
t = 100 * t(f) = 202[ms]
size ~ 1 MiB
and overall transfer speed ~= 5[MiB/s], that is 100 times slower than one large file. The shorter files would yield proportionally lower total transfer speeds.

Predicting which file will be read next is not trivial at all and we'd be delighted to see the community input on possible algorithms and/or patches.

We do have an idea how to deal with very short files, in order of 100 bytes that would cut the latency to about 1/4 or better, but it is not applicable for longer files. 
===========================

**Comment by 4Dolio** [comment]
Would it be possible to perhaps "pack" large quantities of small objects into 64MB chunk files?

Offload the metadata information and contents for such directories entirely to the chunk, perhaps?

**Comment by pkali** [comment]
We have been digging this idea in detail, but it seems to be futile - the main reason is that files do change sizes and this makes the whole idea impractical - we get  fragmentation, issues with files not fitting the space and needing to be moved around, etc.
Much simpler solution is using filesystem in a file that takes care of all of these issues with, possibly, a similar overhead.

**Comment by borkd** [comment]
A cheap trick would be to store first 64k on any object on the fastest label you have (RAM, NVMe, SSD). Yes, that would add some transactions, but it's been battle tested

**Comment by pkali** [comment]
I got pretty excited in the idea for a sec, but unfortunately it would not help - the issue is not transfer speed, but latency. This possibly works miracles in a single machine setup, but here we have 
- client contacts with master
- master replies with chunk locations to client
- client contacts with chunkserver(s) to get the data
And this affects the performance for small files the most. Even the fastest storage will not help here.

The idea for extremely small files might be keeping them on master(s) in metadata storage what would definitely lower the latency.
Quite substantial effort and would not help with files > 127 bytes and very small files are not that numerous.

**Comment by 4Dolio** [comment]
Ya. I use loopbacks, but in doing so you lose the ability for multiple clients to share access to the same set of data.  So I can only use this trick if the data is archival and multiple clients get their own snapshot copy to access read only, or only one client needs access.  Otherwise the loopback client needs to act as a service for multiple clients, which adds more layers.

Would be happy to learn some new tricks to allow multiple clients to access loopbacked fileaystems more natively paralleled..

**Comment by borkd** [comment]
@pkali - Run "file" against a billion files, then repeat 10-100 times.. your grand-grandkids can report the results. Ruthless minimization of latency, where it makes sense, should be merely a milestone to reach maximum aggregate throughput, but not the end goal in itself.

**Comment by borkd** [comment]
@4Dolio - feature/bug creep makes layered r/w loopback usage too exciting. How do you limit blast radius from accidental undergoal or corruption? There is no replication goal to say "I don't care if I loose a couple million of highly replicated files/chunks here or there, but don't loose this multi-terabyte R/W loopback file which contains billions of files and keeps a wayward app happy", is there?

**Comment by pkali** [comment]
Well, you can set goals per file, so one can set goal 1 for disposable files and sth safer for the behemoth container file.

**Comment by 4Dolio** [comment]
I do not worry about loosing chunks. I use minimum of goal 2 and a healthy cluster should never or very rarely drop a chunk.  So I have not encountered any problems with any loopback or iSCSI filesystems.

**Comment by borkd** [comment]
@4Dolio - Claiming the practice should be universally safe because one [has yet to encounter a problem](https://en.wikipedia.org/wiki/Black_swan_theory) may be a disservice to the community. Distributed systems with strange topologies are rarely predictable, and for multi-DC / geo-replicated setups my counter was higher than "never". Given the github project tags tout fault-tolerance, geo-distribution, etc I'd rather label this practice as **could get lucky, proceed at your own risk**.

**Comment by 4Dolio** [comment]
Are you doing geo-rep? That is in and of itself more difficult to tune properly.

Out of the box on a pair of local raspberry pi should be stable. Scaled up to enterprise m.2 is just all that much faster.

Given you have a stable cluster which does not have chunk faults under normal use... Doing loopback/iSCSI should be just as stable.

Yes, even I have a severely overloaded and bloated cluster in production right now.. I need to solve that instability, but even in this case i would expect loopbacks to still work.

If you are loosing chunks, that is not normal.  You should address that first.

**Comment by 4Dolio** [comment]
I did not say it was universally stable. I think i said, given your cluster is stable....

**Comment by uDude** [comment]
Where is the read ahead code?  One of the problems I have with commodity disks is the read ahead hardware cache of the disks is small.  For example WD Red 3/4T commodity only have 64 MB cache in the first place --- i.e., one chunk.  So the software cache code in Lizard is still relevant where we clearly want to cache multiple chunks at least for read purposes.

Predictive techniques are interesting with the chunk abstraction, rather than the disk at the block layer.  Clearly with size(f) > size(chunk) we are trying to reference multiple chunks in parallel and caching may not provide much help unless parallelism is less than the number of necessary chunks.  In any event...

is hdd_readahead.(cc|h) in chunkserver the right location?
